### PR TITLE
Descriptors/dataset

### PR DESCRIPTION
**Pull request by humanpose1**
The whole pipeline for processing 3D match to get fragments.
It downloads the dataset (it is heavy), it contains RGBD frames and the pose and the intrinsic params
Then it creates fragment using [this code](https://github.com/andyzeng/tsdf-fusion-python).
Then it precomputes matches between two fragments (if the overlap is bigger than a parameter).

Still things to do:

- [x] -Hydra conf file.
- [x] Add a Fragment Dataset to train like FCGF (Maybe later).
- [x]  Add a tracker
- [x] pointnet example
- [ ]  dataset for testing
- [ ] saving features
- [ ] script to compute recall
- [ ] script to registrate scene using RANSAC

### CODE REVIEW

File: src/datasets/registration/fragment3dmatch.py [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+from torch_geometric.data import Dataset, Data, download_url, extract_zip
+from torch_geometric.data import DataLoader
+
+from src.datasets.base_dataset import BaseDataset
+from src.datasets.registration.utils import rgbd2fragment
+
+log = logging.getLogger(__name__)
+
+
+def to_list(x):
```
Comment (tchaton) [inline src/datasets/registration/fragment3dmatch.py]: Those helpers should be put within the utils, if there isn't duplicates.

File: src/datasets/registration/fragment3dmatch.py [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    res = []
+    with open(filename, 'r') as f:
+        res = f.readlines()
+    return res
+
+
+class General3DMatch(Dataset):
+
+    list_urls_train = get_urls(osp.join('urls', 'url_train.txt'))
+    list_urls_train_small = get_urls(osp.join('urls', 'url_train_small.txt'))
```
Comment (tchaton) [inline src/datasets/registration/fragment3dmatch.py]: Maybe url_{}.txt could be automatically downloaded.

File: src/datasets/registration/fragment3dmatch.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+
+def get_urls(filename):
+    res = []
+    with open(filename, 'r') as f:
+        res = f.readlines()
+    return res
+
+
+class General3DMatch(Dataset):
```
Comment (tchaton) [inline src/datasets/registration/fragment3dmatch.py]: Maybe, all 3DMatch dataset should be within the same file, using the new API for dataset loading.

class: {file_name}.{dataset_class_name}

File: src/datasets/registration/patch3dmatch.py [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+
+        if(self.transform is not None):
+            data_source = self.transform(data_source)
+            data_target = self.transform(data_target)
+        batch = Batch.from_data_list([data_source, data_target])
+        batch.pair = batch.batch
+        batch.batch = None
+        return batch
+
+class Patch3DMatchDataset(BaseDataset):
```
Comment (tchaton) [inline src/datasets/registration/patch3dmatch.py]: Great !

File: src/datasets/registration/fragment3dmatch.py (hunk 2 above)
Comment (tchaton) [reply src/datasets/registration/fragment3dmatch.py]: Instead of being part of the project. @nicolas-chaulet, what are you thoughts ?

File: src/datasets/registration/fragment3dmatch.py (hunk 3 above)
Comment (humanpose1) [reply src/datasets/registration/fragment3dmatch.py]: I will change the name of the file and adopt the convention.
General3DMatch must not be used directly, that's why it hasn't a dataloader.
It is just a class that download and process the data.
Patch3DMatch or Fragment3DMatch must be used instead.

File: src/datasets/registration/patch3dmatch.py [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+            radius_patch=dataset_opt.radius_patch,
+            num_frame_per_fragment=dataset_opt.num_frame_per_fragment,
+            max_dist_overlap=dataset_opt.max_dist_overlap,
+            min_overlap_ratio=dataset_opt.min_overlap_ratio,
+            tsdf_voxel_size=dataset_opt.tsdf_voxel_size,
+            depth_thresh=dataset_opt.depth_thresh,
+            pre_transform=pre_transform,
+            transform=test_transform)
+
+        self._create_dataloaders(train_dataset, test_dataset)
```
Comment (nicolas-chaulet) [inline src/datasets/registration/patch3dmatch.py]: We are in the process of changing the way datasets and dataloader get created -> See #166. You now simply need to set a `test_dataset` and `train_dataset` members to your dataset object. The train.py scripts takes care of creating the data loaders

File: src/datasets/registration/patch3dmatch.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+        Factory method for the tracker
+
+        Arguments:
+            task {str} -- task description
+            dataset {[type]}
+            wandb_log - Log using weight and biases
+        Returns:
+            [BaseTracker] -- tracker
+        """
+        return PatchRegistrationTracker(dataset, task, wandb_opt.log,
```
Comment (nicolas-chaulet) [inline src/datasets/registration/patch3dmatch.py]: This has changed as well. The wandb and tensorboard options are booleans

File: src/datasets/registration/utils.py [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+        pcd, color = extract_pcd(depth, intrinsic, color_image)
+        pcd = pcd.dot(trans[:3, :3].T) + trans[:3, 3]
+        return pcd, color
+    else:
+        pcd = extract_pcd(depth, intrinsic)
+        pcd = pcd.dot(trans[:3, :3].T) + trans[:3, 3]
+        return pcd
+
+
+def rgbd2fragment_rough(list_path_img,
```
Comment (nicolas-chaulet) [inline src/datasets/registration/utils.py]: Are those functions yours? If not please reference the source.

File: src/metrics/registration_accuracy.py [hunk 8]
```diff
@@ -0,0 +1,18 @@
+import numpy as np
+from sklearn.neighbors import NearestNeighbors
+
+
+def compute_accuracy(embedded_ref_features, embedded_val_features):
+
+    """
+    evaluate accuracy for a batch
```
Comment (nicolas-chaulet) [inline src/metrics/registration_accuracy.py]: Could you please document the input? What are they supposed to contain?

File: src/metrics/registration_tracker.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+        self._acc = compute_accuracy(outputs[::2], outputs[1::2])
+        loss_dict = model.get_current_losses()
+        self._loss = loss_dict["loss"]
+        self._loss_reg = loss_dict["loss_reg"]
+
+    def get_metrics(self, verbose=False) -> Dict[str, float]:
+        """ Returns a dictionnary of all metrics and losses being tracked
+        """
+        metrics = super().get_metrics(verbose)
+        metrics["{}_loss".format(self._stage)] = self._loss
```
Comment (nicolas-chaulet) [inline src/metrics/registration_tracker.py]: You don't need to add the losses, they are part of the `super()` call

File: src/models/base_architectures/backbone.py [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+    """
+    create a backbone-based generator:
+    This is simply an encoder
+    (can be used in classification, regression, metric learning and so one)
+    """
+
+    def _save_sampling_and_search(self, down_conv):
+        sampler = getattr(down_conv, "sampler", None)
+        if is_list(sampler):
+            self._sampling_and_search_dict["sampler"] += sampler
```
Comment (nicolas-chaulet) [inline src/models/base_architectures/backbone.py]: `_sampling_and_search_dict` has been renamed to `_spatial_ops_dict`. Refer to teh Unet as an example

File: src/datasets/registration/fusion.py [hunk 11]
```diff
@@ -0,0 +1,392 @@
+# Copyright (c) 2018 Andy Zeng
+# Taken from https://github.com/andyzeng/tsdf-fusion-python
+
+import numpy as np
+
+from numba import njit, prange
```
Comment (nicolas-chaulet) [inline src/datasets/registration/fusion.py]: Could you please add numba to the poetry dependencies? (poetry add numba)

File: src/datasets/registration/utils.py (hunk 7 above)
Comment (humanpose1) [reply src/datasets/registration/utils.py]: rgbd2fragment_rough is mine (simply concatenate point cloud frame rgbd frame without intelligent fusion). but the fragment are not beautiful with this function.
for rgbd2fragment_fine, I use the code from andy zeng (I have adapted it a bit).

File: src/core/losses/__init__.py [hunk 12]
```diff
@@ -0,0 +1,3 @@
+from .losses import *
+from pytorch_metric_learning.miners import *
```
[the remark this answers is not in the document]
Comment (humanpose1) [reply src/core/losses/__init__.py]: I wanted to do something like instantiate_transforms for the losses and the miners so the user can directly choose the loss from the yaml.
Now it's normally cleaner than before

File: src/datasets/registration/fusion.py (hunk 11 above)
Comment (humanpose1) [reply src/datasets/registration/fusion.py]: I'm sorry. I've had some problems with poetry. Actually, I'm simply using an anaconda env and I've installed the package manually.
We need to add numba, pycuda and pytorch_metric_learning.

File: src/datasets/registration/general3dmatch_dataset.py [hunk 13]
```diff
@@ -0,0 +1,229 @@
+import logging
```
[the remark this answers is not in the document]
Comment (humanpose1) [reply src/datasets/registration/general3dmatch_dataset.py]: I can download a list in the website of 3dmatch [here](http://3dmatch.cs.princeton.edu/), they even have a bash script to download the dataset. But the training dataset and the testing dataset are not split (they also have a split.txt). 
I have made the url.txt manually...

[the remark this answers is not in the document]
Comment (humanpose1) [reply src/datasets/registration/general3dmatch_dataset.py]: also, the training dataset and the testing dataset are different. In the test dataset, we must download precomputed fragments (to have the same data as other papers for the testing).

File: src/datasets/registration/fusion.py (hunk 11 above)
Comment (tchaton) [reply src/datasets/registration/fusion.py]: Did you do that : https://deeppointcloud-benchmarks.readthedocs.io/en/latest/src/installation/

File: src/datasets/registration/general3dmatch_dataset.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (nicolas-chaulet) [reply src/datasets/registration/general3dmatch_dataset.py]: So, if I understand correctly you'd rather keep the txt file as is in the repo?

[the remark this answers is not in the document]
Comment (humanpose1) [reply src/datasets/registration/general3dmatch_dataset.py]: For now, yes even if it's not very elegant... but I will solve this problem later.
